### PR TITLE
Issue #8045: Add COLON token support to NoWhitespaceBefore

### LINKED ISSUE #8045: Add COLON token support to NoWhitespaceBefore
https://checkstyle.sourceforge.io/config_whitespace.html#NoWhitespaceBefore

```
cat TestClass.java
class TestClass {
	{
                mylabel :  // expected violation
		switch(1) {
			case 1 : // expected violation
				break;
			default : // expected violation
				break;
		}
	}
}

$ cat conf.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
       <module name="NoWhitespaceBefore"/>
    </module>
</module>
/var/tmp $ java $RUN_LOCALE -jar checkstyle-8.31-all.jar -c config.xml TestClass.java
Starting audit...
Audit done.
```
https://checkstyle.sourceforge.io/config_whitespace.html#NoWhitespaceBefore_Properties
Currently list of tokens has no support for COLON, it should be added to report violations at least for labels, `case` and `default`, may be there are some other tokens

### PR DESCRIPTION
**Pull request by wilcoln**
Resolve #8045
Diff report : https://wilcoln.github.io/checkstyle-reports/8045/diff/index.html

### CODE REVIEW

**Comment by wilcoln** [comment]
@strkkk, please review.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCheck.java [hunk 1]
```diff
@@ -201,6 +240,21 @@ public void visitToken(DetailAST ast) {
...
+     * @return true if ast is a COLON and is child of LITERAL_CASE or LITERAL_DEFAULT.
+     */
+    private static boolean isColonOfCaseOrDefault(DetailAST colonAst) {
+        boolean result = false;
+        final DetailAST parent = colonAst.getParent();
+        if (parent.getType() == TokenTypes.LITERAL_CASE
+                || parent.getType() == TokenTypes.LITERAL_DEFAULT) {
+            result = true;
+        }
+        return result;
```
Comment (baratali) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCheck.java:255]:
```java
        boolean result = false;
        final DetailAST parent = colonAst.getParent();
        if (parent.getType() == TokenTypes.LITERAL_CASE
                || parent.getType() == TokenTypes.LITERAL_DEFAULT) {
            result = true;
        }
        return result;
```
---->
```java
        final DetailAST parent = colonAst.getParent();
        return parent.getType() == TokenTypes.LITERAL_CASE
                || parent.getType() == TokenTypes.LITERAL_DEFAULT;
```

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespacebefore/InputNoWhitespaceBeforeLabelAndSwitch.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package com.puppycrawl.tools.checkstyle.checks.whitespace.nowhitespacebefore;
+
+class InputNoWhitespaceBeforeLabelAndSwitch
```
Comment (baratali) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespacebefore/InputNoWhitespaceBeforeLabelAndSwitch.java:3]: Please add other cases where COLON can be used. For example, Ternary operator (`boolean ? true : false`), maybe something else.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCheck.java [hunk 3]
```diff
@@ -52,7 +52,11 @@
  * <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#POST_DEC">
  * POST_DEC</a>,
  * <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#ELLIPSIS">
- * ELLIPSIS</a>.
+ * ELLIPSIS</a>,
+ * <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#COLON">
+ * COLON</a>,
```
Comment (baratali) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCheck.java:57]: Please to the docs that the check warns only on colons from cases, default and labels. Or opposite - explain what colons the Check does not check.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespacebefore/InputNoWhitespaceBeforeLabelAndSwitch.java (hunk 2 above)
Comment (wilcoln) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespacebefore/InputNoWhitespaceBeforeLabelAndSwitch.java:3]: Other input files like `InputNoWhitespaceBeforeDefault` already cover this.
If you look at the `testDefault` it runs the check for all tokens, including `COLON`.
Rest assured no false positive is being reported.

Comment (wilcoln) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespacebefore/InputNoWhitespaceBeforeLabelAndSwitch.java:3]: Done, I've added it anyway :)

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCheck.java (hunk 1 above)
Comment (wilcoln) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCheck.java]: Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCheck.java (hunk 3 above)
Comment (wilcoln) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCheck.java:57]: Done.

**Comment by baratali** [comment]
@wilcoln , since this PR changes Check's logic, can you please prepare a diff-report to verify there are no new false-positives/-negatives?

**Comment by wilcoln** [comment]
Of course, I'm uploading it asap.

**Comment by wilcoln** [comment]
@baratali, diff report has been uploaded.

**Comment by rnveach** [comment]
@wilcoln @baratali This issue has not been approved, and I marked my concerns in the issue.

**Comment by wilcoln** [comment]
@rnveach I've opened https://github.com/checkstyle/checkstyle/issues/8124 in response to your comments in https://github.com/checkstyle/checkstyle/issues/8045

**Pull request closed without merging**